flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java

```diff
@@ -107,12 +107,18 @@ protected void reconcileSpecChange(
 
         JobState currentJobState = lastReconciledSpec.getJob().getState();
         JobState desiredJobState = currentDeploySpec.getJob().getState();
+        // check upgrade mode compatibility if deployment was ever stable
+        // an initially unsuccessful deployment can be reverted
+        var hasStableSpec =
+                resource.getStatus().getReconciliationStatus().getLastStableSpec() != null;
```

Comment:
I think this check is too weak. You might have completed checkpoints already at this point (if the job went into a restart loop after some checkpoints but before the operator considered it to be stable).

Also as we improve stability conditions it might become even weaker in the future